drivers/spi/spi_eos_s3.c

```diff
@@ -0,0 +1,353 @@
+}
+
+static int spi_eos_s3_init(const struct device *dev)
+{
+	int ret;
+	const struct spi_eos_s3_config *config = dev->config;
+
+	ret = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (ret) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static inline int spi_eos_s3_next_tx(struct spi_eos_s3_data *data)
+{
+	int tx_frame;
+
+	if (spi_context_tx_buf_on(&data->ctx)) {
+		tx_frame = UNALIGNED_GET((uint8_t *)(data->ctx.tx_buf));
+	} else {
+		tx_frame = NO_TX_DATA;
+	}
+
+	return tx_frame;
+}
+
+static int spi_eos_s3_shift_frames(SPI_TypeDef *regs,
+				   struct spi_eos_s3_data *data)
+{
+	int tx_frame;
+	int rx_frame;
+
+	/* Do not send data when continuous RX mode is selected */
+	if (data->mode != RX_MODE) {
+		tx_frame = spi_eos_s3_next_tx(data);
+		/* Do not send redundant data */
+		if (tx_frame != NO_TX_DATA) {
+			spi_eos_s3_tx(regs, (uint8_t)tx_frame);
+			spi_context_update_tx(&data->ctx, 1, 1);
+		}
+	}
+
+	rx_frame = spi_eos_s3_rx(regs);
+
+	if (rx_frame == NO_RX_DATA)
+		return 0;
+
+	if (spi_context_rx_buf_on(&data->ctx)) {
+		UNALIGNED_PUT((uint8_t)rx_frame, (uint8_t *)data->ctx.rx_buf);
+	}
+	spi_context_update_rx(&data->ctx, 1, 1);
+
+	return 0;
+}
+
+static bool spi_eos_s3_transfer_ongoing(struct spi_eos_s3_data *data)
+{
+	return spi_context_tx_on(&data->ctx) || spi_context_rx_on(&data->ctx);
+}
+
+static void spi_eos_s3_xfer(const struct device *dev)
+{
+	int ret;
+	struct spi_eos_s3_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	const struct spi_eos_s3_config *cfg = dev->config;
+	SPI_TypeDef *regs = cfg->regs;
+
+	if (data->mode == RX_MODE) {
+		/* Set number of bytes for continuous read */
+		spi_eos_s3_set_nframes(regs, ctx->rx_len);
+		/* Send dummy byte*/
+		spi_eos_s3_tx(regs, 0);
+	}
+
+	do {
+		ret = spi_eos_s3_shift_frames(regs, data);
+	} while (!ret && spi_eos_s3_transfer_ongoing(data));
+
+	spi_eos_s3_finish(regs);
+	spi_context_complete(ctx, dev, 0);
+}
+
+/* API Functions */
+static int spi_eos_s3_transceive(const struct device *dev,
+			  const struct spi_config *config,
+			  const struct spi_buf_set *tx_bufs,
+			  const struct spi_buf_set *rx_bufs)
+{
+	const struct spi_eos_s3_config *cfg = dev->config;
+	struct spi_eos_s3_data *data = dev->data;
+	size_t tx_count = 0;
+	size_t rx_count = 0;
+	const struct spi_buf *tx = NULL;
+	const struct spi_buf *rx = NULL;
+	SPI_TypeDef *regs = cfg->regs;
+	int ret = 0;
+
+	data->ctx.config = config;
+
+	if (tx_bufs) {
+		tx = tx_bufs->buffers;
+		tx_count = tx_bufs->count;
+	}
+
+	if (rx_bufs) {
+		rx = rx_bufs->buffers;
+		rx_count = rx_bufs->count;
+	}
+
+	/* Find out what mode is ordered */
+	if (tx_count != 0 && rx_count != 0) {
+		/* TXRX */
+		data->mode = TXRX_MODE;
+	} else if (tx_count != 0) {
+		/* TX */
+		data->mode = TX_MODE;
+	} else {
+		/* RX */
+		data->mode = RX_MODE;
+	}
+	/* EEPROM READ not supported */
+
+	ret = spi_eos_s3_configure(regs, config, data->mode);
+
+	if (ret) {
+		spi_context_release(&data->ctx, ret);
```

Comment:
Btw, you need to call spi_context_lock() at the beginning to this transceive function